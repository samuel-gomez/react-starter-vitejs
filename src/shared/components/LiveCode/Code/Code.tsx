import { createId } from '@axa-fr/react-toolkit-core';
import { Highlight, Prism as defautProps } from 'prism-react-renderer/dist/index';
import { LiveProvider } from 'react-live';

import ClipBoard from '../ClipBoard';

type LiveProviderProps = Parameters<typeof LiveProvider>[0];

type TCode = {
  hideComponent?: boolean;
  title?: string;
  icon?: string;
  theme: LiveProviderProps['theme'];
  code: string;
};

const ariaLabel = 'af-accessibility-code' as const;

const Code = ({ theme, code }: TCode) => (
  <Highlight {...defautProps} theme={theme} code={code} language="jsx">
    {({ className, style, tokens, getLineProps, getTokenProps }) => (
      <>
        <ClipBoard content={code} />
        <pre aria-label={ariaLabel} className={className} style={{ ...style, padding: '1rem', fontSize: '0.8rem', position: 'relative' }}>
          {tokens.map((line, i) => {
            const keyLine = createId();
            return (
              <div key={keyLine} {...getLineProps({ line, key: keyLine })}>
                <span className="token-line__number">{i + 1}</span>
                <span className="token-line__content">
                  {line.map(token => {
                    const key = createId();
                    return <span key={key} {...getTokenProps({ token, key })} />;
                  })}
                </span>
              </div>
            );
          })}
        </pre>
      </>
    )}
  </Highlight>
);

export default Code;
