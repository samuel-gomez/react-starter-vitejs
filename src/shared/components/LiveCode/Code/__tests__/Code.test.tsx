import { render } from '@testing-library/react';
import { themes } from 'prism-react-renderer/dist/index';
import Code from '../Code';

describe('<Code />', () => {
  it('Render <Code /> with theme, code', () => {
    const { asFragment } = render(<Code theme={themes.vsDark} code="test" />);

    expect(asFragment()).toMatchSnapshot();
  });
});
