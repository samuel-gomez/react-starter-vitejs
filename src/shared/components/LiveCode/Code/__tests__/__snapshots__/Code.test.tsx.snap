// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Code /> > Render <Code /> with theme, code 1`] = `
<DocumentFragment>
  <button
    aria-label="Copy to clipboard"
    class="af-btn--circle af-clipboard"
    type="button"
  >
    <i
      class="glyphicon glyphicon-copy"
    />
    <span
      class="af-clipboard__message"
    >
      Copied !
    </span>
  </button>
  <pre
    aria-label="af-accessibility-code"
    class="prism-code language-jsx"
    style="color: rgb(156, 220, 254); background-color: rgb(30, 30, 30); padding: 1rem; font-size: 0.8rem; position: relative;"
  >
    <div
      class="token-line"
      style="color: rgb(156, 220, 254);"
    >
      <span
        class="token-line__number"
      >
        1
      </span>
      <span
        class="token-line__content"
      >
        <span
          class="token plain"
        >
          test
        </span>
      </span>
    </div>
  </pre>
</DocumentFragment>
`;
